### DISCUSSION TITLE
[next 13] Using canonical in generateMetadata

### DISCUSSION THREAD

**Discussion by akiletour** [vercel/next.js#47627] [Help]
### Summary

How to define the canonical url of each page with generateMetadata() in NextJS 13

### Additional information

```js
export async function generateMetadata() {
  const DYNAMIC_URL = ????
  return {
    alternates: {
      canonical: DYNAMIC_URL,
    },
  }
}
```

**Comment by ciruz** [comment]
You can access the `params` and `searchParams` in [`generateMetadata`](https://beta.nextjs.org/docs/api-reference/metadata#generatemetadata).

If you have this structure:

```
app\[slug]\page.tsx
```

Your `generateMetadata` can look like this:

```javascript
// /app/[slug]/page.tsx
export async function generateMetadata({ params, searchParams }) {
  console.log({ params, searchParams });
  return {  };
}
```

If you open following URL `example.com/foobar?test=1`, the server will `console.log` this:

`{ params: { slug: 'foobar' }, searchParams: { test: '1' } }`

So then you can return a canonical with your base URL + the `slug` and maybe some of the `searchParams`.
